biz.aQute.bnd/src/aQute/bnd/main/RemoteCommand.java

```diff
@@ -109,6 +111,57 @@ public void event(Event e) throws Exception {
 		agent = launcher.getAgent();
 	}
 
+	@Description("List the bundles already installed in remote framework")
+	@Arguments(arg = {})
+	interface ListOptions extends Options {
+	}
+
+	public void _list(ListOptions options) throws Exception {
+		List<BundleDTO> installedBundles = agent.getBundles();
+		installedBundles.forEach(this::prettyPrintBundleDTO);
+	}
+
+	@Description("Communicate with the remote agent to install or update bundle")
+	interface InstallOptions extends Options {
+		@Description("Specify the location identifier of the bundle")
+		String location();
+
+		@Description("Specify the file")
+		String file();
+	}
+
+	public void _install(InstallOptions options) throws Exception {
+		String location = options.location();
+		String file = options.file();
+		File f = checkFile(file);
+		if (f != null) {
+			BundleDTO dto = agent.install(location, Files.readAllBytes(f.toPath()));
+			prettyPrintBundleDTO(dto);
+		}
+	}
+
+	private File checkFile(String file) {
+		try {
+			File f = new File(file);
+			Jar jar = new Jar(f);
+			// check if the jar is a bundle
+			if (jar.getBsn() != null) {
+				return f;
+			}
+		} catch (Exception e) {
+			error("Error occurred while parsing the specified file");
+		}
+		return null;
+	}
+
+	private void prettyPrintBundleDTO(BundleDTO dto) {
+		bnd.out.println("==================================================");
```

Comment:
this is very verbose since you also use it to print out the bundles in list. In the list command, I think a table is more appropriate with one bundle per line? I think a simple one-line string would suffice. For example in the standard syntax: <bsn>-<version>[<id>]?

In the list version I think it would be nice to see:

* bsn
* version
* location
* startlevel
* last update
* status

Reply:
We cannot do for now since `BundleDTO` doesn't comprise `location`. To include start level, we need to use `BundleStartLevelDTO` too. As discussed earlier, we will proceed without these properties now. Later on, we could try to use `aQute.bnd.runtime.facade.FrameworkFacade.XBundleDTO`.